nemo/collections/asr/parts/submodules/aed_decoding/aed_batched_streaming.py

```diff
@@ -0,0 +1,562 @@
+                + self.state.prev_encoder_shift  # we need to add the real frame position in the audio signal
+            )
+
+            self.state.decoding_step += input_ids.size(-1)
+
+            # check for hallucinations
+            if self.decoding_cfg.hallucinations_detector:
+                hallucination_mask = self.detect_hallucinations(
+                    self.state.pred_tokens_ids, self.state.batch_idxs, self.state.current_context_lengths
+                )
+                if torch.any(hallucination_mask):
+                    self.state.active_samples *= torch.logical_not(hallucination_mask)
+                    self.state.active_samples_inner_loop *= torch.logical_not(hallucination_mask)
+
+            # disable samples with maximum context length
+            samples_with_max_context_length = (
+                self.state.current_context_lengths == self.state.max_generation_length - 1
+            )
+            if torch.any(samples_with_max_context_length * self.state.active_samples):
+                logging.info("!!! maximum context length reached !!!")
+                self.state.active_samples *= torch.logical_not(samples_with_max_context_length)
+                self.state.active_samples_inner_loop *= torch.logical_not(samples_with_max_context_length)
+
+            # zero out decoder_mems_list for non active samples
+            # TODO batched decoding works wrong if first token was EOS for one of the samples
+            if torch.any(torch.logical_not(self.state.active_samples_inner_loop)):
+                for j in range(len(decoder_mems_list)):
+                    decoder_mems_list[j][:, -1] *= self.state.active_samples_inner_loop.unsqueeze(-1)
+
+            self.state.decoder_mems_list = decoder_mems_list
+            self.state.current_context_lengths += self.state.active_samples_inner_loop
+            # TODO model does not predicts any real tokens in the case of first EOS prediction (rare case for batched decoding)
+            input_ids = self.state.pred_tokens_ids[
+                self.state.batch_idxs, self.state.current_context_lengths - 1
+            ].unsqueeze(-1)
+
+            # limit number of steps per inner loop if not end of speech
+            if self.state.max_tokens_per_alignatt_step is not None:
+                self.state.steps_per_inner_loop += self.state.active_samples_inner_loop
+                disable_samples_mask = self.state.steps_per_inner_loop >= self.state.max_tokens_per_alignatt_step
+                disable_samples_mask *= torch.logical_not(self.state.is_last_chunk_batch)
+                self.state.active_samples_inner_loop *= torch.logical_not(disable_samples_mask)
+
+            if not torch.any(self.state.active_samples_inner_loop):
+                break
+
+    def detect_hallucinations(self, pred_tokens_ids, batch_idxs, current_context_lengths):
+
+        # we need to have at least 8 tokens to run hallucinations detector
+        if self.state.decoding_step < 8:
```

Comment:
Do we have any guarantee that on decoding step 8 we have enough tokens for all samples in the batch?
What if for the first utterance we have 8 tokens, but the last utterance in the batch contains 0 tokens (just silence)?